File: src/ArenaGS/Utilities/Direction.cs
using System;
using System.Collections.Generic;
using ArenaGS.Utilities;

namespace ArenaGS.Utilities
{
	public enum Direction
	{
		None, North, Northeast, East, Southeast, South, Southwest, West, Northwest
	}

	public static class Directions
	{
		public static readonly Direction[] All = new Direction [] { Direction.North, Direction.Northeast, Direction.East, Direction.Southeast,
				Direction.South, Direction.Southwest, Direction.West, Direction.Northwest};
	}

	public static class PointDirectionExtensions
	{
		public static Point InDirection (this Point initial, Direction direction)
		{
			switch (direction)
			{
				case Direction.None:
					return new Point (initial.X, initial.Y);
				case Direction.North:
					return new Point (initial.X, initial.Y - 1);
				case Direction.South:
					return new Point (initial.X, initial.Y + 1);
				case Direction.West:
					return new Point (initial.X - 1, initial.Y);
				case Direction.East:
					return new Point (initial.X + 1, initial.Y);
				case Direction.Northeast:
					return new Point (initial.X + 1, initial.Y - 1);
				case Direction.Northwest:
					return new Point (initial.X - 1, initial.Y - 1);
				case Direction.Southeast:
					return new Point (initial.X + 1, initial.Y + 1);
				case Direction.Southwest:
					return new Point (initial.X - 1, initial.Y + 1);
				default:
					throw new NotImplementedException ();
			}
		}

		public static Direction DirectionTo (this Point initial, Point end)
		{
			int x = end.X - initial.X;
			int y = end.Y - initial.Y;
			if (x > 1)
				x = 1;
			if (x < -1)
				x = -1;
			if (y > 1)
				y = 1;
			if (y < -1)
				y = -1;
			return ConvertPositionDeltaToDirection (x, y);
		}

		static Direction ConvertPositionDeltaToDirection (int deltaX, int deltaY)
		{
			if (deltaX == 1)
			{
				if (deltaY == 1)
					return Direction.Southeast;
				else if (deltaY == -1)
					return Direction.Northeast;
				else
					return Direction.East;
			}
			else if (deltaX == -1)
			{
				if (deltaY == 1)
					return Direction.Southwest;
				else if (deltaY == -1)
					return Direction.Northwest;
				else
					return Direction.West;
			}
			else
			{
				if (deltaY == 1)
					return Direction.South;
				else if (deltaY == -1)
					return Direction.North;
				else
					return Direction.None;
			}
		}

		public static int LatticeDistance (this Point point1, Point point2)
		{
			return Math.Abs (point1.X - point2.X) + Math.Abs (point1.Y - point2.Y);
		}

		public static double NormalDistance (this Point point1, Point point2)
		{
			return Math.Sqrt (Math.Pow (point1.X - point2.X, 2) + Math.Pow (point1.Y - point2.Y, 2));
		}
	}
}

File: src/clash/direction.rs
use crate::atlas::Point;

#[derive(Clone, Copy)]
pub enum Direction {
    None,
    North,
    NorthEast,
    East,
    SouthEast,
    South,
    SouthWest,
    West,
    NorthWest,
}

fn clamp_to_one(x: i32) -> i32 {
    if x > 1 {
        1
    } else if x < -1 {
        -1
    } else {
        x
    }
}

impl Direction {
    pub fn from_two_points(initial: &Point, end: &Point) -> Direction {
        let x = clamp_to_one(end.x as i32 - initial.x as i32);
        let y = clamp_to_one(end.y as i32 - initial.y as i32);

        Direction::from_delta(x, y)
    }

    pub fn from_delta(delta_x: i32, delta_y: i32) -> Direction {
        if delta_x == 1 {
            if delta_y == 1 {
                Direction::SouthEast
            } else if delta_y == -1 {
                Direction::NorthEast
            } else {
                Direction::East
            }
        } else if delta_x == -1 {
            if delta_y == 1 {
                Direction::SouthWest
            } else if delta_y == -1 {
                Direction::NorthWest
            } else {
                Direction::West
            }
        } else {
            if delta_y == 1 {
                Direction::South
            } else if delta_y == -1 {
                Direction::North
            } else {
                Direction::None
            }
        }
    }

    pub fn opposite(&self) -> Direction {
        match self {
            Direction::North => Direction::South,
            Direction::NorthEast => Direction::SouthWest,
            Direction::East => Direction::West,
            Direction::SouthEast => Direction::NorthWest,
            Direction::South => Direction::North,
            Direction::SouthWest => Direction::NorthEast,
            Direction::West => Direction::East,
            Direction::NorthWest => Direction::SouthEast,
            Direction::None => Direction::None,
        }
    }

    pub fn point_in_direction(&self, point: Point) -> Point {
        match self {
            Direction::North => Point::init(point.x, point.y - 1),
            Direction::NorthEast => Point::init(point.x + 1, point.y - 1),
            Direction::East => Point::init(point.x + 1, point.y),
            Direction::SouthEast => Point::init(point.x + 1, point.y + 1),
            Direction::South => Point::init(point.x, point.y + 1),
            Direction::SouthWest => Point::init(point.x - 1, point.y + 1),
            Direction::West => Point::init(point.x - 1, point.y),
            Direction::NorthWest => Point::init(point.x - 1, point.y - 1),
            Direction::None => point,
        }
    }
}
